libflux/src/semantic/walk.rs

```diff
@@ -0,0 +1,1284 @@
+                        Node::FloatLit(ref mut n) => n.typ = MonoType::Var(Tvar(1234)),
+                        Node::StringLit(ref mut n) => n.typ = MonoType::Var(Tvar(1234)),
+                        Node::DurationLit(ref mut n) => n.typ = MonoType::Var(Tvar(1234)),
+                        Node::UintLit(ref mut n) => n.typ = MonoType::Var(Tvar(1234)),
+                        Node::BooleanLit(ref mut n) => n.typ = MonoType::Var(Tvar(1234)),
+                        Node::DateTimeLit(ref mut n) => n.typ = MonoType::Var(Tvar(1234)),
+                        Node::RegexpLit(ref mut n) => n.typ = MonoType::Var(Tvar(1234)),
+                        _ => (),
+                    };
+                },
+                Rc::new(RefCell::new(Node::Package(&mut pkg))),
+            );
+            // now assert that every type is the invalid one
+            let mut v = TypeCollector::new();
+            walk(&mut v, Rc::new(RefCell::new(Node::Package(&mut pkg))));
+            let types = v.types;
+            assert!(types.len() > 0);
+            for tvar in types {
+                assert_eq!(tvar, Tvar(1234));
+            }
+        }
+    }
+
+    mod nesting {
+        use super::*;
+        use crate::ast::Operator::AdditionOperator;
+
+        // NestingCounter counts the number of nested Blocks found while walking.
+        struct NestingCounter {
+            count: u8,
+        }
+
+        impl Visitor for NestingCounter {
+            fn visit(&mut self, node: Rc<RefCell<Node>>) -> bool {
+                match *node.borrow() {
+                    Node::Block(_) => self.count += 1,
+                    _ => (),
+                }
+                true
+            }
+        }
+
+        #[test]
+        fn test_nesting_count() {
+            let mut pkg = compile(
+                r#"
+f = () => {
+    // 1
+    return () => {
+        // 2
+        return () => {
+            // 3
+            1 + 1
+            // 4
+            return 1
+        }
+    }
+}
+
+g = () => {
+    // 5
+    return 2
+}
+
+f()()()
+g()
+"#,
+            );
+            let mut v = NestingCounter { count: 0 };
+            walk(&mut v, Rc::new(RefCell::new(Node::Package(&mut pkg))));
+            assert_eq!(v.count, 5);
+        }
+
+        // RepeatedPlusChecker checks if there is more than one addition in the same Function Scope.
+        // If so, it errors.
+        struct RepeatedPlusChecker {
+            // A stack.
+            plus: Vec<bool>,
+            err: String,
+        }
+
+        impl RepeatedPlusChecker {
+            fn new() -> RepeatedPlusChecker {
+                RepeatedPlusChecker {
+                    plus: vec![false],
+                    err: "".to_string(),
+                }
+            }
+        }
+
+        impl Visitor for RepeatedPlusChecker {
+            fn visit(&mut self, node: Rc<RefCell<Node>>) -> bool {
+                match *node.borrow() {
+                    Node::FunctionExpr(_) => {
```

Comment:
now this must match `FunctionExpr` cause we have multiple lines in a function body and every line is a `Block`